File: CodeSmells/dispensables/data-class.ts
import {v4 as uuidv4} from 'uuid'

// Code smell: Data Class [Clase de datos]. UserRecord expone datos públicos con poco o ningún comportamiento,
// empujando toda la lógica a otras clases y fomentando modelos de dominio anémicos.

// Ejercicio: Requiere verificación de email y reglas de dominio (p. ej., solo company.com).

// Necesitarás tocar múltiples servicios y lugares que manipulan UserRecord,
// mostrando cómo separar el comportamiento de los datos hace que cambios simples se dispersen ampliamente.

export class UserRecord {
  constructor(
    public id: string,
    public name: string,
    public email: string,
    public createdAt: Date,
  ) {
  }
}

class UserService {
  createUser(name: string, email: string): UserRecord {
    if (!email.includes('@')) {
      throw new Error('Invalid email')
    }

    return new UserRecord(uuidv4(), name, email, new Date())
  }

  updateUserEmail(user: UserRecord, newEmail: string): void {
    if (!newEmail.includes('@')) {
      throw new Error('Invalid email')
    }
    user.email = newEmail
  }
}

class UserReportGenerator {
  generateUserSummary(user: UserRecord): string {
    return `User ${user.name} (${user.email}) created on ${user.createdAt.toLocaleDateString()}`
  }
}

// Example usage orchestrating behavior in separate services rather than the data class itself
export function demoDataClass(): string {
  const service = new UserService()
  const report = new UserReportGenerator()
  const user = service.createUser('Lina', 'lina@example.com')
  service.updateUserEmail(user, 'lina+news@example.com')
  return report.generateUserSummary(user)
}

File: CodeSmells/dispensables/data-class.cs
using System;

// Code smell: Data Class [Clase de datos]. UserRecord expone datos públicos con poco o ningún comportamiento,
// empujando toda la lógica a otras clases y fomentando modelos de dominio anémicos.

// Ejercicio: Requiere verificación de email y reglas de dominio (p. ej., solo company.com).

// Necesitarás tocar múltiples servicios y lugares que manipulan UserRecord,
// mostrando cómo separar el comportamiento de los datos hace que cambios simples se dispersen ampliamente.

namespace CodeSmells.Dispensables
{
    public class UserRecord
    {
        public string id;
        public string name;
        public string email;
        public DateTime createdAt;

        public UserRecord(string id, string name, string email, DateTime createdAt)
        {
            this.id = id;
            this.name = name;
            this.email = email;
            this.createdAt = createdAt;
        }
    }

    class UserService
    {
        public UserRecord CreateUser(string name, string email)
        {
            if (!email.Contains("@"))
            {
                throw new Exception("Invalid email");
            }

            return new UserRecord(Guid.NewGuid().ToString(), name, email, DateTime.Now);
        }

        public void UpdateUserEmail(UserRecord user, string newEmail)
        {
            if (!newEmail.Contains("@"))
            {
                throw new Exception("Invalid email");
            }
            user.email = newEmail;
        }
    }

    class UserReportGenerator
    {
        public string GenerateUserSummary(UserRecord user)
        {
            return $"User {user.name} ({user.email}) created on {user.createdAt.ToShortDateString()}";
        }
    }

    // Example usage orchestrating behavior in separate services rather than the data class itself
    public static class DataClassDemo
    {
        public static string DemoDataClass()
        {
            var service = new UserService();
            var report = new UserReportGenerator();
            var user = service.CreateUser("Lina", "lina@example.com");
            service.UpdateUserEmail(user, "lina+news@example.com");
            return report.GenerateUserSummary(user);
        }
    }
}
